test/30_genbranch.bats

```diff
@@ -57,3 +57,78 @@ setup() {
+
+  git pile genpatches -m "First pile commit"
+  run --separate-stderr git pile genbranch -i
+  [ "${lines[0]}" = "Applying: 1st commit after baseline" ]
+  [ "${lines[1]}" = "Applying: 2nd commit after baseline" ]
+  [ "${lines[2]}" = "Applying: 3rd commit after baseline" ]
+  [ "${lines[3]}" = "Applying: 4th commit after baseline" ]
+  [ ${#lines[@]} -eq 4 ]
+
+  tip_first_series=$(git rev-parse HEAD)
+
+  git reset --hard HEAD~2
+
+  echo "pile iii" > j.txt && git add j.txt && git commit -m "iii commit after baseline"
+  echo "pile iv" > j.txt && git add j.txt && git commit -m "iv commit after baseline"
+
+  git pile genpatches -m "Second pile commit"
+  run --separate-stderr git pile genbranch -i
+  [ "${lines[0]}" = "Applying: iii commit after baseline" ]
+  [ "${lines[1]}" = "Applying: iv commit after baseline" ]
+  [ ${#lines[@]} -eq 2 ]
+
+  git -C "$(git config pile.dir)" reset --hard HEAD~1
+  run --separate-stderr git pile genbranch -i
+  [ ${#lines} -eq 0 ]
+  [ $(git rev-parse HEAD) == "$tip_first_series" ]
+
+  # Cases below should trigger a "full genbranch"
+
+  run --separate-stderr git -c pile.genbranch-user-name="pile bot 2" pile genbranch -i
+  [ "${lines[0]}" = "Applying: 1st commit after baseline" ]
+  [ "${lines[1]}" = "Applying: 2nd commit after baseline" ]
+  [ "${lines[2]}" = "Applying: 3rd commit after baseline" ]
+  [ "${lines[3]}" = "Applying: 4th commit after baseline" ]
+  [ ${#lines[@]} -eq 4 ]
+
+  run --separate-stderr git pile genbranch -i --fix-whitespace
+  [ "${lines[0]}" = "Applying: 1st commit after baseline" ]
+  [ "${lines[1]}" = "Applying: 2nd commit after baseline" ]
+  [ "${lines[2]}" = "Applying: 3rd commit after baseline" ]
+  [ "${lines[3]}" = "Applying: 4th commit after baseline" ]
+  [ ${#lines[@]} -eq 4 ]
+  [[ "$stderr" = *"warning: Caching disabled because of non-default genbranch operation: using option --fix-whitespace"* ]]
+
+  run --separate-stderr git -c pile.genbranch-use-cache=false pile genbranch -i
+  [ "${lines[0]}" = "Applying: 1st commit after baseline" ]
+  [ "${lines[1]}" = "Applying: 2nd commit after baseline" ]
+  [ "${lines[2]}" = "Applying: 3rd commit after baseline" ]
+  [ "${lines[3]}" = "Applying: 4th commit after baseline" ]
+  [ ${#lines[@]} -eq 4 ]
+
+  run --separate-stderr git pile genbranch -i --no-cache
+  [ "${lines[0]}" = "Applying: 1st commit after baseline" ]
+  [ "${lines[1]}" = "Applying: 2nd commit after baseline" ]
+  [ "${lines[2]}" = "Applying: 3rd commit after baseline" ]
+  [ "${lines[3]}" = "Applying: 4th commit after baseline" ]
+  [ ${#lines[@]} -eq 4 ]
+
+  run --separate-stderr git -c pile.genbranch-cache-path=other-file.pickle pile genbranch -i
+  [ "${lines[0]}" = "Applying: 1st commit after baseline" ]
+  [ "${lines[1]}" = "Applying: 2nd commit after baseline" ]
+  [ "${lines[2]}" = "Applying: 3rd commit after baseline" ]
+  [ "${lines[3]}" = "Applying: 4th commit after baseline" ]
+  [ ${#lines[@]} -eq 4 ]
```

Comment:
What I'm not seeing and not sure it works with the current structures is:

1) genbranch -i to have the full cache
2) someone else updates the pile/result branch, replacing "2nd commit after baseline" with "ii commit after baseline"
3) we reset pile/result-branch to this already created result-branch/pile (that we got from the remote)

Here it means we don't have the cache for anything after 1st commit

4) We apply a patch changing "3rd commit after baseline" to "iii commit after the baseline".


Can we also add a test for reordering commits without going through pile commands?

Reply:
> What I'm not seeing and not sure it works with the current structures is:
> 
>     1. genbranch -i to have the full cache
> 
>     2. someone else updates the pile/result branch, replacing "2nd commit after baseline" with "ii commit after baseline"
> 
>     3. we reset pile/result-branch to this already created result-branch/pile (that we got from the remote)
> 
> 
> Here it means we don't have the cache for anything after 1st commit
> 
>     4. We apply a patch changing "3rd commit after baseline" to "iii commit after the baseline".
> 

Just force-pushed adding this scenario. Could you check if I captured it correctly, please?

> 
> Can we also add a test for reordering commits without going through pile commands?

Not sure if I follow. Do you mean the following? (1) genbranch with the four commits; (2) reorder them just with "normal" git commands, without touching the pile branch; (3) genbranch again

Reply:
yeah, that covers it, thanks

Reply:
> > Can we also add a test for reordering commits without going through pile commands?
> 
> Not sure if I follow. Do you mean the following? (1) genbranch with the four commits; (2) reorder them just with "normal" git commands, without touching the pile branch; (3) genbranch again

I am assuming you were talking about yet another test case here, right? I'm not sure if your "yes" response above is for this as well or only for the test I had pushed.

Reply:
> I am assuming you were talking about yet another test case here, right?

yes. And my "yes" was to "(1) genbranch with the four commits; (2) reorder them just with "normal" git commands, without touching the pile branch; (3) genbranch again".

while playing with this with some manual tests everything seems to be working. There may be issues with the way to check the tests though:
1) If the branch had already been generated before, it will simply have no output because it's reusing older commits. This we can kind of control since we create a new repo from scratch.
2) If git-gc triggers in the middle, we would lose some objects and fail the test

I think this is all theoretical problems only that may not be worth changing. One thing we could do in future if we do see problems like this is to change the test comparison:  What we want to know is if genbranch generates the exact same branch as when called with --no-cache. So we could simply generate that branch, invalidate the cache, setup the test, do the test we want, compare HEAD.

Reply:
> There may be issues with the way to check the tests though:
> 
>     1. If the branch had already been generated before, it will simply have no output because it's reusing older commits. This we can kind of control since we create a new repo from scratch.
> 
>     2. If git-gc triggers in the middle, we would lose some objects and fail the test

For (1), as you noted, I think we are fine here, since we are testing things in a controlled environment: `$BATS_TEST_TMPDIR` will be unique for each test.

For (2), because of the size of the test repository, I do not think we would reach the threshold for triggering automatic gc. If we want to be sure, we could disable it with `git config gc.auto 0`.

> (...) One thing we could do in future if we do see problems like this is to change the test comparison: What we want to know is if genbranch generates the exact same branch as when called with --no-cache.

I do not think that would be enough. Even without using the cache `genbranch` could also generate exactly the same branch (when using commiter-date-is-author-date). I think would need to check if `genbranch` skipped the first `k` patches, `k` being the number of patches that already have the corresponding commit objects in the repository.

Reply:
> I do not think that would be enough. Even without using the cache genbranch could also generate exactly the same branch (when using commiter-date-is-author-date).

yeah, that's why it was added. But the point here is that with or without cache, the same thing is generated. If the faster path (with cache) generates the same `git rev-parse HEAD` I'm sitting on, I'm pretty happy. I know a bug in the cache implementation didn't trigger it to generate something wrong. I prefer testing that the final output makes sense because then the implementation can change underneath without breaking tests.

Reply:
Force-pushed with some changes in the tests. Now there is a `run_genbranch` function that checks if the history matches the one generated without using cache.